src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryContext.java

```diff
@@ -604,6 +605,50 @@ public void watchTree(Object path)
     }
   }
 
+  @StarlarkMethod(
+      name = "repo_metadata",
+      doc =
+          """
+          Constructs an opaque object that can be returned from the repo rule's implementation \
+          function to provide metadata about its reproducibility.
+          """,
+      parameters = {
+        @Param(
+            name = "reproducible",
+            defaultValue = "False",
+            doc =
+                """
+                States that this repo can be reproducibly refetched; that is, if it were fetched \
+                another time with exactly the same input attributes, repo rule definition, watched \
+                files and environment variables, etc., then exactly the same output would be \
+                produced. This property needs to hold even if other untracked conditions change, \
+                such as information from the internet, the path of the workspace root, output from \
+                running arbitrary executables, etc. If set to True, this allows the fetched repo \
+                contents to be cached across workspaces. \
+                <p>Note that setting this to True does not guarantee caching in the repo contents \
+                cache; for example, local repo rules are never cached.
+                """,
+            positional = false,
+            named = true),
+        @Param(
+            name = "attrs_for_reproducibility",
+            defaultValue = "{}",
+            doc =
+                """
+                If <code>reproducible</code> is False, this can be specified to tell Bazel which \
+                attributes of the original repo rule to change to make it reproducible.
+                """,
+            positional = false,
+            named = true)
+      })
+  public RepoMetadata repoMetadata(boolean reproducible, Dict<?, ?> attrs) throws EvalException {
+    if (reproducible && !attrs.isEmpty()) {
+      throw Starlark.errorf(
+          "attrs_for_reproducibility can only be specified if reproducible is False");
```

Comment:
In the future, this failure case would exactly fit the use case of https://github.com/bazelbuild/bazel/issues/24777#issuecomment-2674494229.

Reply:
ack.